### PR TITLE
[SPARK-11530] [MLLIB] Return eigenvalues with PCA model

### PR DESCRIPTION
**Pull request by srowen**
Add `computePrincipalComponentsAndVariance` to also compute PCA's explained variance.

CC @mengxr

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala [hunk 1]
```diff
@@ -379,25 +379,43 @@ class RowMatrix @Since("1.0.0") (
    * Note that this cannot be computed on matrices with more than 65535 columns.
    *
    * @param k number of top principal components.
-   * @return a matrix of size n-by-k, whose columns are principal components
+   * @return a matrix of size n-by-k, whose columns are principal components, and
+   * a vector of values which indicate how much variance each principal component
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala]: It is not very clear from the doc whether we return the absolute variance explained or the proportions. How about `a vector of proportions of variance explained by each principal component` and change the method name to `computePrincipalComponentsAndExplainedVariance`?

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala]: Definitely, sounds good.

**Comment by srowen** [comment]
@mengxr what do you think about this one? it's adding some weight to the API; that's my main concern. Having the explained variance is nice though.

**Comment by srowen** [comment]
@mengxr I'm going to push this for 1.7, not 1.6 -- thoughts?

**Comment by srowen** [comment]
Merged to master

**Comment by jkbradley** [comment]
@srowen I didn't see this PR before, but there will need to be a follow-up in order to make the model save/load backwards compatible.  We can check the saved metadata for the Spark version, and then expect the explainedVariance val to be there only if the version is > 1.6.

I just created [https://issues.apache.org/jira/browse/SPARK-12349] for this.

**Comment by srowen** [comment]
@jkbradley ah, even if it's an experimental API? OK, seems simple enough. It's no problem to ignore the extra new column in old versions. If the new column isn't present, it's a little funny to figure out what the explained variance should be. Probably empty or null is the best that can be done -- or is it best to fail anyway?

**Comment by jkbradley** [comment]
@srowen I agree we didn't really promise to have backwards compatibility yet, but it'd be great if we could maintain it.

Good question about API.  Rather than being fancy about throwing errors, I'd say just set it to be an empty Vector and note in the .load Scala docs that it will be empty for models saved using Spark 1.6.

**Comment by srowen** [comment]
@jkbradley great, have a look at https://github.com/apache/spark/pull/10327

**Pull request closed without merging**